crates/core/src/live_socket/navigation/ffi.rs

```diff
@@ -0,0 +1,285 @@
+            info,
+        }
+    }
+
+    pub fn new_from_reload(dest: NavHistoryEntry, info: Option<Vec<u8>>) -> NavEvent {
+        NavEvent::new(NavEventType::Reload, dest.clone(), dest.into(), info)
+    }
+
+    /// Create a new nav event from the details of a [NavCtx::traverse_to] event
+    pub fn new_from_traverse(
+        new_dest: NavHistoryEntry,
+        old_dest: NavHistoryEntry,
+        info: Option<Vec<u8>>,
+    ) -> NavEvent {
+        NavEvent::new(NavEventType::Traverse, new_dest, old_dest.into(), info)
+    }
+
+    /// Create a new nav event from the details of a [NavCtx::navigate] event
+    pub fn new_from_navigate(
+        new_dest: NavHistoryEntry,
+        old_dest: Option<NavHistoryEntry>,
+        opts: NavOptions,
+    ) -> NavEvent {
+        let event = match opts.action {
+            Some(NavAction::Replace) => NavEventType::Replace,
+            _ => NavEventType::Push,
+        };
+
+        NavEvent::new(event, new_dest, old_dest, opts.extra_event_info)
+    }
+
+    /// Create a new nav event from the details of a [NavCtx::back] event,
+    /// passing info into the event handler closure.
+    pub fn new_from_forward(
+        new_dest: NavHistoryEntry,
+        old_dest: Option<NavHistoryEntry>,
+        info: Option<Vec<u8>>,
+    ) -> NavEvent {
+        NavEvent::new(NavEventType::Push, new_dest, old_dest, info)
+    }
+
+    /// Create a new nav event from the details of a [NavCtx::back] event,
+    /// passing info into the event handler closure.
+    pub fn new_from_back(
+        new_dest: NavHistoryEntry,
+        old_dest: NavHistoryEntry,
+        info: Option<Vec<u8>>,
+    ) -> NavEvent {
+        NavEvent::new(NavEventType::Push, new_dest, Some(old_dest), info)
+    }
+}
+
+use crate::live_socket::socket::SessionData;
+
+use super::{super::error::LiveSocketError, LiveSocket, NavCtx};
+
+impl LiveSocket {
+    /// Tries to navigate to the current item in the NavCtx.
+    /// changing state in one fell swoop if initialilization succeeds
+    async fn try_nav(&self) -> Result<(), LiveSocketError> {
+        let current = self
+            .current()
+            .ok_or(LiveSocketError::NavigationImpossible)?;
+
+        let url = Url::parse(&current.url)?;
+
+        let format = self.session_data.lock().expect("poison").format.clone();
+        let options = self
+            .session_data
+            .lock()
+            .expect("poison")
+            .connect_opts
+            .clone();
+
+        let session_data = SessionData::request(&url, &format, options).await?;
+        let websocket_url = session_data.get_live_socket_url()?;
+        let socket = Socket::spawn(websocket_url, Some(session_data.cookies.clone())).await?;
+
+        self.socket()
+            .disconnect()
+            .await
+            .map_err(|_| LiveSocketError::DisconnectionError)?;
+
+        *self.socket.lock().expect("poison") = socket;
+        *self.session_data.lock().expect("poison") = session_data;
```

Comment:
Any reason you're doing `lock().expect("poison")` rathen than `?`. I know that it's an unhandle-able error but it is nice to bubble up.